pkg/object/udpproxy/session.go

```diff
@@ -0,0 +1,224 @@
+/*
+ * Copyright (c) 2017, MegaEase
+ * All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package udpproxy
+
+import (
+	"fmt"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/megaease/easegress/pkg/logger"
+	"github.com/megaease/easegress/pkg/util/fasttime"
+	"github.com/megaease/easegress/pkg/util/iobufferpool"
+	"github.com/megaease/easegress/pkg/util/timerpool"
+)
+
+type (
+	session struct {
+		clientAddr        *net.UDPAddr
+		serverAddr        string
+		serverConn        net.Conn
+		clientIdleTimeout time.Duration
+		serverIdleTimeout time.Duration
+		writeBuf          chan *iobufferpool.Packet
+
+		stopped      bool
+		stopChan     chan struct{}
+		listenerStop chan struct{}
+		onClose      func()
+
+		mu sync.Mutex
+	}
+)
+
+func newSession(clientAddr *net.UDPAddr, serverAddr string, serverConn net.Conn,
+	listenerStop chan struct{}, onClose func(),
+	clientIdleTimeout, serverIdleTimeout time.Duration) *session {
+	s := session{
+		serverAddr:        serverAddr,
+		clientAddr:        clientAddr,
+		serverConn:        serverConn,
+		serverIdleTimeout: serverIdleTimeout,
+		clientIdleTimeout: clientIdleTimeout,
+		writeBuf:          make(chan *iobufferpool.Packet, 512),
+
+		stopped:      false,
+		stopChan:     make(chan struct{}),
+		listenerStop: listenerStop,
+		onClose:      onClose,
+	}
+
+	go func() {
```

Comment:
nit: This `newSession` constructor is a bit long. I would consider refactoring lines 67-120 to a private function, for example `startSession` etc and then here call `go s.startSession()`.

Reply:
Good idea.
I will refractor it.